Allow to disable users registration and/or login

leio10 opened the pull request:
#### :tophat: What? Why?
I'm working in a Decidim instance that should only allow users to access throught an OmniAuth integration. So, I've added a `users_registration_mode` setting to `Decidim::Core` to allow to customize this.

This setting has three posible values: `:enabled`, `:existing` and `:disabled`. `:enabled` is the default value and it will keep Decidim working as now. On the other hand, `:disabled` disables completely user registration and sign in with email and password. This does not affect OmniAuth integration, this kind of registration will  work even when regular registration is disabled.

The last possible value, `:existing`, allow users to sign in but not to sign up. I think that this could be helpful when working with a fixed list of users. As said before, this case doesn't affect OmniAuth integration, so I don't know if this scenario would have sense if there are OmniAuth integrations enabled. I've added this option because I think that could be helpful and it doesn't need any extra work, but I don't really need it. If you think that wouldn't be useful or could be confusing I can remove it and convert the setting to a boolean `users_registration_enabled` variable.

#### :pushpin: Related Issues

#### :clipboard: Subtasks
- [x] Disable Devise controller actions based on the setting
- [x] Hide links based on the setting
- [x] Add tests
- [x] Added default setting to application initializer template
- [x] Add `CHANGELOG` entry

### :camera: Screenshots (optional)

- With `:disabled` value:
![imagen](https://user-images.githubusercontent.com/453545/48060796-4d7e6e00-e1bd-11e8-8931-5137aec62ad0.png)
![imagen](https://user-images.githubusercontent.com/453545/48061652-03e35280-e1c0-11e8-9d7d-c58175bbe863.png)
![imagen](https://user-images.githubusercontent.com/453545/48060819-5bcc8a00-e1bd-11e8-9fa1-da7674219fdb.png)

- With `:existing` value:
![imagen](https://user-images.githubusercontent.com/453545/48060796-4d7e6e00-e1bd-11e8-8931-5137aec62ad0.png)
![imagen](https://user-images.githubusercontent.com/453545/48061135-3be99600-e1be-11e8-91e2-ee106d2f9f17.png)
![imagen](https://user-images.githubusercontent.com/453545/48061715-2bd2b600-e1c0-11e8-8e79-4e2410de13ee.png)

MarcReniu commented:
Hello @leio10. This is a very good feature!! But I have to point out that this is not suitable for multitenant installations.

The `users_registration_mode` should be a configuration option in System BO. This means that `Organization` needs a new field to store that value, and all the new login to management and control moved.

leio10 commented:
Hi @MarcReniu, thanks for the feedback!

I can do that, but I'd love to hear some @decidim/product thougths before investing additional efforts on this. :smile_cat:

andreslucena commented:
Hi all!
We love the feature, thanks for the work. 
As a minimum, it should be configurable for multitentats as @mrcasals said. 
I can think at least of an internal instance that we're talking inside the Barcelona City Hall, basically an instance for every city worker. The user registration and login would be used externally through LDAP, so as it is would be perfect. 

On the other hand, for other kind of processes, we need to have more granularity regarding what to enable/disable, for instance to disallow user registration but allow user login, for small organizations that wants to have a closed census. As far as I know, this have been done for two organizations: Federació Catalana d’Escoltisme i Guiatge (FCEG) and FundAction. Maybe this is another feature, but I think we should support this kind of granularity sooner or later.

leio10 commented:
@andreslucena @MarcReniu Just to confirm, I'm only adding this setting to the system zone, not to the admin zone. I'm guessing that this setting would be changed by the same person that changes the host or the authorizations enabled for an organization.

MarcReniu commented:
@leio10 Yes, this setting should only be available to System admins.

leio10 commented:
Well @decidim/developers, this is ready to be reviewed! (and https://github.com/decidim/decidim/pull/4440 too :smile:)

oriolgual commented on decidim-core/app/models/decidim/organization.rb:
```diff
@@ -22,6 +22,12 @@ class Organization < ApplicationRecord
     has_many :oauth_applications, foreign_key: "decidim_organization_id", class_name: "Decidim::OAuthApplication", inverse_of: :organization, dependent: :destroy
     has_many :hashtags, foreign_key: "decidim_organization_id", class_name: "Decidim::Hashtag", dependent: :destroy
 
+    # Users registration mode. Whether users can register or access the system. Doesn't affect users that access through Omniauth integrations.
```
I'd add some docs at `system` or getting started too.

oriolgual commented on decidim-core/db/migrate/20181108131058_add_users_registration_mode_to_organizations.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddUsersRegistrationModeToOrganizations < ActiveRecord::Migration[5.2]
+  def change
+    add_column :decidim_organizations, :users_registration_mode, :integer, default: 0
```
Why is this an integer?

oriolgual replied:
Shouldn't we update the factories, too?

leio10 commented:
@oriolgual thanks for the feedback!

> I'd add some docs at system or getting started too.

I've added a [new document in the customizations folder](https://github.com/decidim/decidim/blob/7580ebe546f8b14ab54d540c8f67d48823017c90/docs/customization/users_registration_mode.md) to explain this feature. I've also checked if there is any place in the getting started, the checklist or the system docs to add a link to it, but I didn't find anything there related to the organization configuration.
Please, tell me if you know any place where I can add it without messing too much. Anyway, I think that maybe just adding this new document to https://docs.decidim.org/ could be enougth.

> Why is this an integer?

I've reviewed [enum docs](https://api.rubyonrails.org/v5.2.1/classes/ActiveRecord/Enum.html) and enums fields should be integer. Anyway, I've added a not nullable setting to the column, as in [other enum fields in Decidim](https://github.com/decidim/decidim/blob/master/decidim-initiatives/db/migrate/20170906094044_create_decidim_initiatives.rb#L31)

> Shouldn't we update the factories, too?

Updated!

mrcasals commented:
Wohoo, thanks for the PR @leio10! 😄 🎉
